mat64/io.go

```diff
@@ -114,6 +150,69 @@ func (m *Dense) UnmarshalBinary(data []byte) error {
 	return nil
 }
 
+// UnmarshalBinaryFrom decodes the binary form into the receiver and returns
+// the number of bytes read and an error if any.
+// It panics if the receiver is a non-zero Dense matrix.
+//
+// See MarshalBinary for the on-disk layout.
+//
+// Limited checks on the validity of the binary input are performed:
+//  - matrix.ErrShape is returned if the number of rows or columns is negative,
+//  - an error is returned if the resulting Dense matrix is too
+//  big for the current architecture (e.g. a 16GB matrix written by a
+//  64b application and read back from a 32b application.)
+// UnmarshalBinary does not limit the size of the unmarshaled matrix, and so
+// it should not be used on untrusted data.
+func (m *Dense) UnmarshalBinaryFrom(r io.Reader) (int, error) {
+	if !m.isZero() {
+		panic("mat64: unmarshal into non-zero matrix")
+	}
+
+	var (
+		n   int
+		buf [8]byte
+	)
+	nn, err := readFull(r, buf[:])
+	n += nn
+	if err != nil {
+		return n, err
+	}
+	rows := int64(binary.LittleEndian.Uint64(buf[:]))
+
+	nn, err = readFull(r, buf[:])
+	n += nn
+	if err != nil {
+		return n, err
+	}
+	cols := int64(binary.LittleEndian.Uint64(buf[:]))
+	if rows < 0 || cols < 0 {
+		return n, errBadSize
+	}
+
+	size := rows * cols
+	if int(size) < 0 || size > maxLen {
+		return n, errTooBig
+	}
+
+	m.mat.Rows = int(rows)
```

Comment:
is the additional function call worth it?
`reuseAs` makes a few sanity checks that we either already do or know by construction that we don't need.

well, I did a quick bench for Dense w/ and w/o `reuseAs`:

```
$ benchstat no-reuseas.txt ./reuseas.txt 
name                       old time/op    new time/op    delta
UnmarshalDense10-4            262ns ± 1%     269ns ± 4%    ~             (p=0.206 n=5+5)
UnmarshalDense100-4          1.70µs ± 3%    1.71µs ± 8%    ~             (p=0.841 n=5+5)
UnmarshalDense1000-4         15.5µs ± 2%    15.1µs ± 2%  -2.65%          (p=0.016 n=5+5)
UnmarshalDense10000-4         139µs ± 2%     138µs ± 6%    ~             (p=0.222 n=5+5)
UnmarshalFromDense10-4        619ns ± 2%     622ns ± 1%    ~             (p=0.468 n=5+5)
UnmarshalFromDense100-4      4.49µs ± 1%    4.55µs ± 2%    ~             (p=0.095 n=5+5)
UnmarshalFromDense1000-4     42.7µs ± 3%    43.0µs ± 3%    ~             (p=0.841 n=5+5)
UnmarshalFromDense10000-4     419µs ± 3%     426µs ± 4%    ~             (p=0.310 n=5+5)

name                       old alloc/op   new alloc/op   delta
UnmarshalDense10-4            80.0B ± 0%     80.0B ± 0%    ~     (all samples are equal)
UnmarshalDense100-4            896B ± 0%      896B ± 0%    ~     (all samples are equal)
UnmarshalDense1000-4         8.19kB ± 0%    8.19kB ± 0%    ~     (all samples are equal)
UnmarshalDense10000-4        81.9kB ± 0%    81.9kB ± 0%    ~     (all samples are equal)
UnmarshalFromDense10-4        88.0B ± 0%     88.0B ± 0%    ~     (all samples are equal)
UnmarshalFromDense100-4        904B ± 0%      904B ± 0%    ~     (all samples are equal)
UnmarshalFromDense1000-4     8.20kB ± 0%    8.20kB ± 0%    ~     (all samples are equal)
UnmarshalFromDense10000-4    81.9kB ± 0%    81.9kB ± 0%    ~     (all samples are equal)

name                       old allocs/op  new allocs/op  delta
UnmarshalDense10-4             1.00 ± 0%      1.00 ± 0%    ~     (all samples are equal)
UnmarshalDense100-4            1.00 ± 0%      1.00 ± 0%    ~     (all samples are equal)
UnmarshalDense1000-4           1.00 ± 0%      1.00 ± 0%    ~     (all samples are equal)
UnmarshalDense10000-4          1.00 ± 0%      1.00 ± 0%    ~     (all samples are equal)
UnmarshalFromDense10-4         2.00 ± 0%      2.00 ± 0%    ~     (all samples are equal)
UnmarshalFromDense100-4        2.00 ± 0%      2.00 ± 0%    ~     (all samples are equal)
UnmarshalFromDense1000-4       2.00 ± 0%      2.00 ± 0%    ~     (all samples are equal)
UnmarshalFromDense10000-4      2.00 ± 0%      2.00 ± 0%    ~     (all samples are equal)
```

it seems the extra function call isn't impacting too much.
it's definitely in the noise.